### PR TITLE
Integrate Calico network plugin with vagrant provisioner

### PR DESCRIPTION
**Pull request by Symmetric**
I've been working on a prototype integration of [Calico](https://github.com/Metaswitch/calico-docker) networking in Kubernetes. I've used the new [network plugin API](https://github.com/GoogleCloudPlatform/kubernetes/pull/5069) to set up the Pod infra container with Calico networking, and the basic use-case is working.

This adds tenant isolation to Kubernetes, so that pod traffic can be restricted using the Calico distributed firewall. It also means that network traffic between nodes is unencapsulated, since Calico uses an L3 BGP network rather than an L2 overlay network like OVS/flannel.

At this point I'd like to get some feedback on the general approach, to see if there's anything that I should be doing differently.
# Overview

Here's an overview of the structure:
- Add a NETWORK_MODE to chose between openvswitch and calico.
- Update the salt files to install the Calico infrastructure container on each minion.
- Update the salt files to run a new etcd instance on the master, listening on port 6666
  - Needed for the Calico etcd datastore, which can't be run as a pod since we need Calico to be up before we can start pods.
- On pod creation:
  - Record the existing (docker-allocated) pod IP.
  - Delete the bridged veth.
  - Add a new veth pair with the docker-allocated Pod IP, using Calico networking. The node-end of the veth pair is not attached to a vSwitch or docker bridge, instead it is processed by the Linux routing table (i.e. a vRouter).
# Questions

This works fine with direct Pod-Pod communication, but I'm not sure about Services -- in order to firewall a flow properly, we need to see the real source IP when it hits the iptables rules on the destination node. I'm concerned that the Service is going to get in the way of this; it looks like the service proxy is rewriting the source IP of packets to stay in the reverse flow, is that correct?

If so, any suggestions for how we could get the Service proxy to play nice with Calico?

And, some general questions on the architecture described above:
- Should I move the calico-node container initialization from the salt files to the plugin's `setup` step?
  - I use salt config for setting up the calico node, so not sure what's preferable there.

Any other feedback on the design is welcome.

CC: @thockin @lxpollitt
# TODO
- [x] Add `setup` handler in network plugin
- [x] Add `teardown` handler in network plugin
- [x] Set default policy profile in Calico plugin, so that pods have global connectivity by default
- [x] Tidy up OVS/calico flags
- [x] Figure out how Services are going to work (wait for https://github.com/GoogleCloudPlatform/kubernetes/issues/6666 and/or https://github.com/GoogleCloudPlatform/kubernetes/pull/9210)

### CODE REVIEW

**Comment by Symmetric** [comment]
After digging into the kube-proxy a bit more, I'm pretty sure we're not going to be able to do firewalling with the current Kubernetes implemenatation.

If a client sends a packet to a service IP, then the service pod is going to see a request from an IP owned by kube-proxy. This means that we can't do inbound filtering at the pod, since we don't know the client's true source address/port. Outbound filtering at the client's end is unwieldy here, though possible with some extensions; we could watch the API for new services, and install outbound firewall rules blocking access to certain services. But this isn't as effective as inbound filtering at the pod, since it doesn't protect against compromised endpoints on the network.

Looks like we need something like DNS SRV (or multiple A records), as suggested by https://github.com/GoogleCloudPlatform/kubernetes/issues/6666. In that case the client would communicate directly with the pod, and our firewall would work at both the traffic source and destination.

**Comment by Symmetric** [comment]
I signed it! (as Metaswitch Networks).

**Comment by Symmetric** [comment]
cc @spikecurtis

**Comment by lxpollitt** [comment]
Note that I'm not sure why this is labeled "cla: no" since the contributors license agreement has been signed and submitted.

**Comment by jdef** [comment]
Would the proxy still be in the way (appearing as "the client") if everything was done via iptables instead of userspace proxy impl?
- https://github.com/GoogleCloudPlatform/kubernetes/pull/9210
  - https://github.com/GoogleCloudPlatform/kubernetes/issues/3760

**Comment by spikecurtis** [comment]
@jdef #9210 should preserve the actual pod IP as the source when the packets hit the service instances, which would allow our firewall to operate correctly (yay!).

It may also alleviate the difficulty on older kernels of `REDIRECT` versus `DNAT` rules that currently forward traffic to the proxy (`REDIRECT` rules don't work for Calico if the proxy is not bound to a specific IP address on the kubelet).  I'd need to check with @Symmetric to be sure about what testing he's done in this area.

**Comment by BenTheElder** [comment]
Note that when looking at #9210 the iptables rules as generated now may not be entirely correct, I'll be working to finish checking that once vagrant is working again. We also haven't selected a target iptables minimum version yet.

Any comments or suggestions on the rules used etc would be appreciated, I'd like to get this right the first time, and I'd like to make it play nice with the work here.

**Comment by Symmetric** [comment]
Some fairly major refactoring has just gone in. I've rebased to the latest master, which has required some change to the structure (e.g. how/where we set up the cbr0 bridge). Functionally I think we're good, but there are a couple extraneous bash variables still lying around.

Also, I've moved the plugin code onto the calico-docker repository. It's versioned, so new builds there won't break what's checked in here.

**Comment by thockin** [comment]
I finally took a look at this.  Overall it's in-line with the current codebase, so I can't complain, but I will anyway.  The salt is becoming a real tangle of options that are impossible to ensure work in all cases.  As cross through the 1.0 gate we need to think hard about how we're going to a) make this sort of thing more self-contained, b) make this testable, c) offload the testing to sub-owners to ensure (e.g.) Calico keeps working without our own test machines running e2e N-times on every PR.

**Comment by Symmetric** [comment]
@thockin thanks for taking a look. I agree that the salt files are pretty tangled; it took a lot more effort than expected to get this working. I have been pondering what it would look like in an ideal world, and I have a few ideas for possible improvements.

I think that it would be worth exploring the possibility of pulling out the provisioning providers in kubernetes into separate repos, rather than checking them into the main k8s repo. The Vagrant provider model is a good one; if structured correctly, externally-hosted providers could be importable by the main k8s cluster tool. That way a standard UI for cluster operations could be maintained, without requiring all provider code to live in the k8s repo.

I'm up for spending some time kicking ideas around and working on kube-provisioning 2.0, and refactoring the Calico provider to the New Way. I imagine there will be a very wide audience for that discussion though.

Would you be happy to take this PR, and address the more fundamental questions of provisioning plugins once the dust settles from 1.0?

**Comment by piosz** [comment]
Is this PR still active?

**Comment by caseydavenport** [comment]
I believe @Symmetric was still hoping to get this in, but was waiting for some more feedback / direction from @thockin.  It looks like this needs a rebase, and @Symmetric will likely be inactive until Monday.  We can wait to see what he has to say on it, or close it for now and we can re-open when he gets back.

File: cluster/vagrant/provision-master.sh [hunk 1]
```diff
@@ -71,8 +71,15 @@ done
 echo "127.0.0.1 localhost" >> /etc/hosts # enables cmds like 'kubectl get pods' on master.
 echo "$MASTER_IP $MASTER_NAME" >> /etc/hosts
 
-# Configure the openvswitch network
-provision-network
+if [ $NETWORK_MODE != "calico" ]; then
```
Comment (thockin) [inline cluster/vagrant/provision-master.sh]: I would make this explicitly == "openvswitch"

File: cluster/vagrant/provision-master.sh [hunk 2]
```diff
@@ -71,8 +71,15 @@ done
...
+  # Configure the openvswitch network
+  echo "Provisioning openvswitch network"
+  provision-network
+else
```
Comment (thockin) [inline cluster/vagrant/provision-master.sh]: else if "calico"

File: cluster/vagrant/provision-master.sh [hunk 3]
```diff
@@ -71,8 +71,15 @@ done
...
+  # Configure the cbr for Calico networking.
+  echo "Provisioning cbr0 for Calico"
+  provision-network-calico
+fi
```
Comment (thockin) [inline cluster/vagrant/provision-master.sh]: else abort

**Comment by thockin** [comment]
I'm OK with this.  My complaints from before stand, but they are not your fault.

@derekwaynecarr for final LGTM for vagrant

**Comment by Symmetric** [comment]
Thanks for taking a look.

Review markups applied in 4c580f4, want me to squash now or wait for final ack?

**Comment by Symmetric** [comment]
@derekwaynecarr -- any chance you can take a look at this? Would be great to minimize the pain of rebasing and get this in...

Just been resized as XS, so shouldn't take much of your time. Cheers!

**Comment by kelseyhightower** [comment]
@Symmetric Hello! I promise if you rebase this again I'll spend the time to finish up the review. Also thanks for taking the time to put together this PR.

**Comment by Symmetric** [comment]
Cheers Kelsey, I'll rebase this week, hopefully later today.

**Comment by Symmetric** [comment]
@kelseyhightower this turned out to be more involved than I thought, but it's working now; PTAL.

@caseydavenport will be taking this over the finish line.

Cheers!

**Comment by derekwaynecarr** [comment]
@Symmetric @caseydavenport - please poke the PR when its ready again.

also please update with a note in here:
https://github.com/kubernetes/kubernetes/tree/master/docs/getting-started-guides

that you will provide future support for calico based network deployment issues for the affected platforms.

Thanks!

**Comment by dcbw** [comment]
@Symmetric this and https://github.com/kubernetes/kubernetes/pull/18795 are going to clash a bit, so whichever of us merges first we'll have to sort that out.

**Comment by lxpollitt** [comment]
@caseydavenport: see @dcbw's comment above and please coordinate with him

**Comment by caseydavenport** [comment]
@dcbw - Yep.  I think this PR is already quite out-of-date as it is.  I'd suggest you go first, and I'll worry about this guy afterwards.

I think we can close this PR, and I can open up another when I get around to rebasing, etc.

**Pull request closed without merging**